pageserver/src/page_cache.rs

```diff
@@ -96,183 +123,460 @@ impl Slot {
 }
 
 struct SlotInner {
-    key: Option<(MapKey, Lsn)>,
+    key: Option<MapKey>,
     buf: &'static mut [u8; PAGE_SZ],
 }
 
 impl SlotInner {
-    fn is_for_key(&self, map_key: &MapKey, lsn: Lsn) -> bool {
-        match self.key.as_ref() {
-            Some((self_map_key, self_lsn)) => self_map_key == map_key && self_lsn == &lsn,
+    fn is_for_key(&self, map_key: &MapKey) -> bool {
+        match &self.key {
+            Some(self_key) => self_key == map_key,
             None => false,
         }
     }
 }
 
 pub struct PageCache {
     // TODO use a more write friendly datastructure
-    map: RwLock<HashMap<MapKey, Vec<Version>>>,
+    materialized_page_map: RwLock<HashMap<MaterializedPageHashKey, Vec<Version>>>,
+    // TODO: If you add support for a new kind of object to cache, add a new
+    // mapping data structure for the that object here.
     slots: Box<[Slot]>,
 
     next_evict_slot: AtomicUsize,
 }
 
-pub struct LookupGuard<'i>(RwLockReadGuard<'i, SlotInner>);
+pub struct PageReadGuard<'i>(RwLockReadGuard<'i, SlotInner>);
 
-impl std::ops::Deref for LookupGuard<'_> {
+impl std::ops::Deref for PageReadGuard<'_> {
     type Target = [u8; PAGE_SZ];
 
     fn deref(&self) -> &Self::Target {
         self.0.buf
     }
 }
 
-impl PageCache {
-    fn new(num_pages: usize) -> Self {
-        // TODO avoid memory leaks
-        let page_buffer = Box::leak(vec![0u8; num_pages * PAGE_SZ].into_boxed_slice());
+pub struct PageWriteGuard<'i> {
+    inner: RwLockWriteGuard<'i, SlotInner>,
+    valid: bool,
+}
```

Comment:
I understand why you'd want to return a guard for reads: the entry might not be used so it's nice to avoid the copy. But why for writes? A copy isn't required (we can just send in a ref)

Reply:
Now I see the use case in #864. Does it makes sense to lock the slot for the duration of the read from disk, vs reading to a temporary buffer and then writing into the cache later?

Reply:
I think it's better to keep the buffer locked. If another thread tries to lock the buffer, it means that the other thread wants to also read or write the same page. It's better for the other thread to sleep until this thread has finished reading in the page. If we didn't keep the buffer locked, then both threads would read the page into a temporary buffer, which is wasteful, but it's also a bit scary. What if both threads wanted to modify the page? Would they both read in the page from disk to the temporary buffer, and perform the modifications on the temporary buffer? You'd need to detect the conflict when you write the page to cache. Or make sure that the modifications are idempotent, so that you never have two threads that would try to do two *different* modifications to the same page.

All in all, simpler and better to just keep the page locked.